include/compiler.hpp

```diff
@@ -0,0 +1,248 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+#pragma once
+
+#include <vector>
+#include <memory>
+#include <mutex>
+#include <variant>
+#include <cstddef>
+
+class Logger
+{
+protected:
+    std::recursive_mutex &m_lock;
+    std::string &m_buildLog;
+
+public:
+    Logger(std::recursive_mutex &lock, std::string& build_log)

```

Comment:
I wish C++ had better indications of using references as opposed to value semantics at calls sites, but I do strongly prefer references here because these values should never be null. Using pointers implies they can potentially be null and should be checked.